src/taco-dependency-installer/test/javaJdkInstaller.ts

```diff
@@ -0,0 +1,344 @@
+            // child_process.exec call will fail for attach dmg, but succeed for detach dmg
+            var timesCalled = 0;
+            childProcessModule.exec = (command: string, optionsOrCallback: nodeFakes.ExecSecondArgument,
+                callback: nodeFakes.Callback = null): nodeFakes.IChildProcess => {
+                var realCallback = <nodeFakes.Callback> (callback || optionsOrCallback);
+                if (timesCalled++ === 0) {
+                    // This is for the attach call
+                    realCallback(new Error("Couldn't attach dmg file"), /*stdout*/ new Buffer("/Volumes/Macintosh HD"), /*stderr*/ new Buffer(""));
+                } else {
+                    // This is for the detach call
+                    realCallback(null, /*stdout*/ new Buffer(""), /*stderr*/ new Buffer(""));
+                }
+                return new nodeFakes.ChildProcess();
+            };
+
+            var expectedTelemetry: TacoUtility.ICommandTelemetryProperties[] = [
+                {
+                    "initialStep.time": { isPii: false, value: "2000" },
+                    step: { isPii: false, value: "initialStep" }
+                },
+                {
+                    "error.description": { isPii: false, value: "ErrorOnChildProcess on attachDmg" },
+                    "install.time": { isPii: false, value: "2000" },
+                    lastStepExecuted: { isPii: false, value: "install" },
+                    step: { isPii: false, value: "install" },
+                    time: { isPii: false, value: "3000" }
+                }
+            ];
+
+            return telemetryGeneratedShouldBe(expectedTelemetry, /Couldn\'t attach dmg file/, done);
+        });
+
+        function fakeExecToTestPackageInstallerFailure(command: string, optionsOrCallback: nodeFakes.ExecSecondArgument,
+            callback: nodeFakes.Callback = null): nodeFakes.IChildProcess {
+            var realCallback = <nodeFakes.Callback> (callback || optionsOrCallback);
+            if (callTime === 0) {
+                // This is for the attach call
+                realCallback(null, /*stdout*/ new Buffer("/Volumes/Macintosh HD"), /*stderr*/ new Buffer(""));
+            } else if (callTime === 1) {
+                // This is for the install package call
+                var couldntRunPackageInstallerError = new Error("Package installer exited with an error");
+                switch (childProcessExecBehavior) {
+                    case ChildProcessExecBehavior.PackageInstallerCommandFailsScenario:
+                        (<any> couldntRunPackageInstallerError).code = 54;
+                        break;
+                    case ChildProcessExecBehavior.PackageInstallerCommandCantBeRunScenario:
+                        break;
+                    case ChildProcessExecBehavior.DetachingDmgFileFailsScenario:
+                        couldntRunPackageInstallerError = null;
+                        break;
```

Comment:
Just for future-proofing the authoring of more test cases, I would add a default case in this switch with some generic error like "Bad test configuration". This goes with my above suggestion of reseting `childProcessExecBehavior` to null before/after each test.